### PR TITLE
Improve semantics of out-of-bounds indices in gather and scatter

### PR DESCRIPTION
**Pull request by hawkinsp**
Previously the behavior of out-of-bounds indexed accesses was not well-defined. Gathers clamped indices to be in-bounds, whereas out-of-bounds scatters were dropped. Additionally, the gradients of out-of-bounds indices were incorrect, since scatter (the transpose of gather) had different semantics to gather.

This PR adds a new `mode` option to `gather` that allows the user to select one of three modes:
* `"clip"`: clamp out-of-bounds indices into range
* `"fill"`: fill out-of-bounds gathered windows with a constant value, typically NaN or 0.
* `"promise_in_bounds"`: preserve the current behavior, which is `"clip"` for gathers.

The PR also adds the same `mode` option to `scatter`. The modes for `scatter` are identical, but `"fill"` is an alias for `"drop"`, meaning "drop out of bounds updates".

The PR gives two main benefits:
* if we change the default for `gather` to `"fill"`, then we can help users catch accidental out-of-bounds indexing. If no `fill_value` is provided, `gather` defaults to one of NaN, or an extremal integer value. This should indicate the presence of out-of-bounds indices.
*  Since we now have both `"clip"` and `"fill"`/`"drop"` modes for both `gather` and `scatter`, we can now define a correct transpose for out-of-bounds indices. The transpose of a `"clip"` gather is a `"clip"` scatter, and the transpose of a `"fill"` gather is a `"drop"` scatter.

We probably need to add the same semantics to `dynamic_slice` and `dynamic_update_slice` in a future PR.

This PR does not yet change the default mode, since that is likely to break users. For now, it is set to `"promise_in_bounds"`, which preserves the current behavior.

### CODE REVIEW

File: jax/_src/lax/lax.py [hunk 1]
```diff
@@ -966,12 +966,45 @@ class GatherDimensionNumbers(NamedTuple):
...
+
+  @staticmethod
+  def from_any(s: Optional[Union[str, 'GatherScatterMode']]):
+    if s == "clip" or s == GatherScatterMode.CLIP:
+      return GatherScatterMode.CLIP
+    if s == "fill" or s == "drop" or s == GatherScatterMode.FILL_OR_DROP:
+      return GatherScatterMode.FILL_OR_DROP
+    if (s is None or s == "promise_in_bounds" or
+        s == GatherScatterMode.PROMISE_IN_BOUNDS):
+      return GatherScatterMode.PROMISE_IN_BOUNDS
```
Comment (shoyer) [inline jax/_src/lax/lax.py:995]: `isintance` on the enum might help simplify this:
```suggestion
    if isinstance(s, GatherScatterMode):
      return s
    if s == "clip":
      return GatherScatterMode.CLIP
    if s == "fill" or s == "drop":
      return GatherScatterMode.FILL_OR_DROP
    if s is None or s == "promise_in_bounds":
      return GatherScatterMode.PROMISE_IN_BOUNDS
```

File: jax/_src/lax/lax.py [hunk 2]
```diff
@@ -966,12 +966,45 @@ class GatherDimensionNumbers(NamedTuple):
   start_index_map: Sequence[int]
 
 
+class GatherScatterMode(IntEnum):
```
Comment (shoyer) [inline jax/_src/lax/lax.py:969]: Is there any particular reason for using `IntEnum` here? Generally `Enum` is a better choice for non-legacy code, which does not need to support integer values.

**Review by shoyer — COMMENTED** [review]
partial review: I did not look at the implemenation, only the inferface

File: jax/_src/lax/lax.py (hunk 1 above)
Comment (hawkinsp) [reply jax/_src/lax/lax.py]: Done.

File: jax/_src/lax/lax.py (hunk 2 above)
Comment (hawkinsp) [reply jax/_src/lax/lax.py]: Changed to `Enum`. Thanks for the tip!

File: jax/_src/lax/lax.py [hunk 3]
```diff
@@ -966,12 +966,46 @@ class GatherDimensionNumbers(NamedTuple):
   start_index_map: Sequence[int]
 
 
+class GatherScatterMode(Enum):
+  # Indices will be clamped to the nearest in-range value, i.e., such that the
+  # entire window to be gathered is in-range.
+  CLIP = 0
```
Comment (mattjj) [inline jax/_src/lax/lax.py:972]: Do you like using `auto`, as in `from enum import auto` and `CLIP = auto()` etc?

(My favorite way to write enums is like namedtuples, as in `GatherScatterMode = enum.Enum('GatherScatterMode', ['CLIP', 'FILL_OR_DROP', 'PROMISE_IN_BOUNDS'])`. But then we don't have space for the comments, or for adding the `from_any` method, so probably not the best option here. Just thought I'd share!)

**Review by mattjj — APPROVED** [review]
Great change, and writeup!

File: docs/jax.ops.rst [hunk 4]
```diff
@@ -21,6 +21,7 @@ follows (where ``idx`` is a NumPy index expression).
 =========================  ===================================================
 Alternate syntax           Equivalent in-place expression
 =========================  ===================================================
+``x.at[idx].get(y)``       ``x[idx]``
```
Comment (soraros) [inline docs/jax.ops.rst:24]: I think this should be `x.at[idx].get()`?

Comment (hawkinsp) [reply docs/jax.ops.rst:24]: Thanks! Yes. #8000 fixes it.
